### PR TITLE
MOM6: +(*)Barotropic improvements and various cleanup

### PR DESCRIPTION
**Pull request by Hallberg-NOAA**
Added a number of options to the barotropic solver improving the algorithms
and made a series of minor improvements and code cleanup, addressing a number
of issues that had been flagged in the code with "###".  By default, all answers
in the test cases are bitwise identical, but there are new entries in the
MOM_parameter_doc files.  The commits in this PR include:

- NOAA-GFDL/MOM6@d7663f4 +Added two new run-time parameters for kappa_shear
- NOAA-GFDL/MOM6@db6cf47 +Added the new parameter BULKML_CONV_MOMENTUM_BUG
- NOAA-GFDL/MOM6@bfa0e15 Removed unused variable in ePBL_column
- NOAA-GFDL/MOM6@a6c4261 +Add dimensional rescaling to Idealized_Hurricane
- NOAA-GFDL/MOM6@a6c4261 +Add dimensional rescaling to Idealized_Hurricane
- NOAA-GFDL/MOM6@e7983f5 Merge branch 'dev/gfdl' into minor_corrections
- NOAA-GFDL/MOM6@01318a8 +Added multiple options for the barotropic solver
- NOAA-GFDL/MOM6@b9a5e7b (*)Rescaled FLUXCONST during get_param
- NOAA-GFDL/MOM6@f6da690 (*)Average Stokes drifts over accumulated depths
- NOAA-GFDL/MOM6@378a02c Started refactoring determine_temperature
- NOAA-GFDL/MOM6@93fc465 +(*)Added runtime parameter HOR_REGRID_2018_ANSWERS
- NOAA-GFDL/MOM6@d20d551 +Added cont_stencil arg to initialize_dyn routines
- NOAA-GFDL/MOM6@6bf9284 +Added MEKE_GEOMETRIC_2018_ANSWERS
- NOAA-GFDL/MOM6@f2e4644 (+)Added the runtime parameter KH_BG_2D_BUG
- NOAA-GFDL/MOM6@765fa37 Added parentheses in calculate_bkgnd_mixing
- NOAA-GFDL/MOM6@12d3aff (*)+Added run-time parameter SURFACE_2018_ANSWERS
- NOAA-GFDL/MOM6@3388236 (+)Corrected units of ADAPT_TIME_RATIO

### CODE REVIEW

**Comment by Hallberg-NOAA** [comment]
This was tested with https://gitlab.gfdl.noaa.gov/ogrp/MOM6/pipelines/9793, which passed with the expected warnings about changing MOM_parameter_doc files.

File: src/user/Idealized_Hurricane.F90 [hunk 1]
```diff
@@ -361,72 +379,87 @@ subroutine idealized_hurricane_wind_profile(CS, absf, YY, XX, UOCN, VOCN, Tx, Ty
...
 
   !  Use a simple drag coefficient as a function of U10 (from Sullivan et al., 2010)
   du10 = sqrt(du**2+dv**2)
-  if (du10.lt.11.) then
+  if (dU10 < 11.0*US%m_s_to_L_T) then
      Cd = 1.2e-3
-  elseif (du10.lt.20.0) then
-     Cd = (0.49 + 0.065*U10)*1.e-3
+  elseif (dU10 < 20.0*US%m_s_to_L_T) then
+     Cd = (0.49 + 0.065*US%L_T_to_m_s*U10)*1.e-3
```
Comment (breichl) [inline src/user/Idealized_Hurricane.F90]: Note this U10 should also have been changed to dU10.

File: src/user/Idealized_Hurricane.F90 [hunk 2]
```diff
@@ -565,35 +612,41 @@ subroutine SCM_idealized_hurricane_wind_forcing(state, forces, day, G, US, CS)
...
-       Cd = 1.2e-3
-    elseif (du10 < 20.) then
-       Cd = (0.49 + 0.065 * U10 )*0.001
+    du10 = sqrt(du**2+dv**2)
+    if (dU10 < 11.0*US%m_s_to_L_T) then
+      Cd = 1.2e-3
+    elseif (dU10 < 20.0*US%m_s_to_L_T) then
+      if (CS%answers_2018) then
+        Cd = (0.49 + 0.065 * US%L_T_to_m_s*U10 )*0.001
+      else ! Brandon, please verify that this line should use du10 instead of U10 -RWH
```
Comment (breichl) [inline src/user/Idealized_Hurricane.F90]: This should be dU10, correct.

File: src/user/Idealized_Hurricane.F90 [hunk 3]
```diff
@@ -565,35 +612,41 @@ subroutine SCM_idealized_hurricane_wind_forcing(state, forces, day, G, US, CS)
...
-       Cd = 1.2e-3
-    elseif (du10 < 20.) then
-       Cd = (0.49 + 0.065 * U10 )*0.001
+    if (dU10 < 11.0*US%m_s_to_L_T) then
+      Cd = 1.2e-3
+    elseif (dU10 < 20.0*US%m_s_to_L_T) then
+      if (CS%answers_2018) then
+        Cd = (0.49 + 0.065 * US%L_T_to_m_s*U10 )*0.001
+      else ! Brandon, please verify that this line should use du10 instead of U10 -RWH
+        Cd = (0.49 + 0.065 * US%L_T_to_m_s*dU10 )*0.001
```
Comment (breichl) [inline src/user/Idealized_Hurricane.F90]: Again, correct to use dU10 here.

**Comment by Hallberg-NOAA** [comment]
The updated PR is being tested with https://gitlab.gfdl.noaa.gov/ogrp/MOM6/pipelines/9811.

File: src/core/MOM.F90 [hunk 4]
```diff
@@ -2933,7 +2967,7 @@ subroutine extract_surface_state(CS, sfc_state)
     !$OMP parallel do default(shared)
     do j=js,je ; do i=is,ie
       ! Convert from gSalt to kgSalt
-      sfc_state%salt_deficit(i,j) = 1000.0 * US%R_to_kg_m3*US%Z_to_m*CS%tv%salt_deficit(i,j)
+      sfc_state%salt_deficit(i,j) = 0.001 * US%R_to_kg_m3*US%Z_to_m*CS%tv%salt_deficit(i,j)
```
Comment (marshallward) [inline src/core/MOM.F90]: This is a bugfix, right?  Is the old expression not needed here?

Comment (marshallward) [reply src/core/MOM.F90]: @Hallberg-NOAA has commented that he has no knowledge of anyone using this, and that the error is so egregious that it's almost certainly unused by anyone, so no need to retain the old behavior.

**Comment by marshallward** [comment]
Tests have passed up to a parameter change; will update MOM6-examples soon.

**Pull request merged**